// @langchain/exa/dist/retrievers.cjs
let _langchain_core_retrievers = require("@langchain/core/retrievers");
let _langchain_core_documents = require("@langchain/core/documents");
//#region src/retrievers.ts
function _getMetadata(result) {
	const newMetadata = { ...result };
	delete newMetadata.text;
	return newMetadata;
}
/**
* @example
* ```typescript
* const retriever = new ExaRetriever({
*   client: new Exa(
*     process.env.EXA_API_KEY,
*     process.env.EXA_BASE_URL,
*   ),
* });
* const docs = await retriever.getRelevantDocuments("hello");
* ```
*/
var ExaRetriever = class extends _langchain_core_retrievers.BaseRetriever {
	static lc_name() {
		return "ExaRetriever";
	}
	lc_namespace = [
		"langchain",
		"retrievers",
		"exa"
	];
	client;
	searchArgs;
	constructor(fields) {
		super(fields);
		this.client = fields.client;
		this.searchArgs = fields.searchArgs;
	}
	async _getRelevantDocuments(query) {
		const res = await this.client.searchAndContents(query, this.searchArgs);
		const documents = [];
		for (const result of res.results) {
			let pageContent;
			if ("text" in result) pageContent = result.text;
			else if ("highlights" in result) pageContent = result.highlights.join("\n\n");
			else pageContent = "No results found.";
			documents.push(new _langchain_core_documents.Document({
				pageContent,
				metadata: _getMetadata(result)
			}));
		}
		return documents;
	}
};
//#endregion
exports.ExaRetriever = ExaRetriever;
exports._getMetadata = _getMetadata;
